File: backend/trip-service/internal/service/constants_service.go
package service

import (
	"context"

	"bus-booking/trip-service/internal/constants"
	"bus-booking/trip-service/internal/model"
)

type ConstantsService interface {
	GetBusConstants(ctx context.Context) (*model.BusConstants, error)
	GetRouteConstants(ctx context.Context) (*model.RouteConstants, error)
	GetTripConstants(ctx context.Context) (*model.TripConstants, error)
	GetSearchFilterConstants(ctx context.Context) (*model.SearchFilterConstants, error)
	GetCities(ctx context.Context) ([]string, error)
	GetAllConstants(ctx context.Context) (*model.ConstantsResponse, error)
}

type ConstantsServiceImpl struct{}

func NewConstantsService() ConstantsService {
	return &ConstantsServiceImpl{}
}

func (s *ConstantsServiceImpl) GetBusConstants(ctx context.Context) (*model.BusConstants, error) {
	seatTypes := make([]model.SeatTypeConstant, 0)
	for _, st := range constants.AllSeatTypes() {
		seatTypes = append(seatTypes, model.SeatTypeConstant{
			Value:           st.String(),
			DisplayName:     st.GetDisplayName(),
			PriceMultiplier: st.GetPriceMultiplier(),
		})
	}

	amenities := make([]model.AmenityConstant, 0)
	for _, a := range constants.AllAmenities {
		amenities = append(amenities, model.AmenityConstant{
			Value:       a.String(),
			DisplayName: a.GetDisplayName(),
		})
	}

	busTypes := make([]model.BusTypeConstant, 0)
	for _, bt := range constants.AllBusTypes() {
		busTypes = append(busTypes, model.BusTypeConstant{
			Value:       bt.String(),
			DisplayName: bt.GetDisplayName(),
		})
	}

	return &model.BusConstants{
		SeatTypes: seatTypes,
		Amenities: amenities,
		BusTypes:  busTypes,
	}, nil
}

func (s *ConstantsServiceImpl) GetRouteConstants(ctx context.Context) (*model.RouteConstants, error) {
	stopTypes := make([]model.StopTypeConstant, 0)
	for _, st := range constants.AllStopTypes() {
		stopTypes = append(stopTypes, model.StopTypeConstant{
			Value:       st.String(),
			DisplayName: st.GetDisplayName(),
		})
	}

	return &model.RouteConstants{
		StopTypes: stopTypes,
	}, nil
}

func (s *ConstantsServiceImpl) GetTripConstants(ctx context.Context) (*model.TripConstants, error) {
	tripStatuses := make([]model.TripStatusConstant, 0)
	for _, ts := range constants.AllTripStatuses() {
		tripStatuses = append(tripStatuses, model.TripStatusConstant{
			Value:       ts.String(),
			DisplayName: ts.GetDisplayName(),
		})
	}

	return &model.TripConstants{
		TripStatuses: tripStatuses,
	}, nil
}

func (s *ConstantsServiceImpl) GetSearchFilterConstants(ctx context.Context) (*model.SearchFilterConstants, error) {
	// Price ranges - single range for user selection
	priceRanges := []model.FilterPriceRange{
		{Min: 0, Max: 1000000},
	}

	// Time slots
	timeSlots := []model.FilterTimeSlot{
		{StartTime: "00:00", EndTime: "06:00", DisplayName: "Sáng sớtm (00:00 - 06:00)"},
		{StartTime: "06:00", EndTime: "12:00", DisplayName: "Ban ngày (06:00 - 12:00)"},
		{StartTime: "12:00", EndTime: "18:00", DisplayName: "Chiều (12:00 - 18:00)"},
		{StartTime: "18:00", EndTime: "24:00", DisplayName: "Tối (18:00 - 24:00)"},
	}

	// Seat types - reuse from bus constants
	seatTypes := make([]model.SeatTypeConstant, 0)
	for _, st := range constants.AllSeatTypes() {
		seatTypes = append(seatTypes, model.SeatTypeConstant{
			Value:           st.String(),
			DisplayName:     st.GetDisplayName(),
			PriceMultiplier: st.GetPriceMultiplier(),
		})
	}

	// Amenities - filter common ones for search
	commonAmenities := []constants.Amenity{
		constants.AmenityWiFi,
		constants.AmenityAC,
		constants.AmenityWater,
		constants.AmenityCharging,
		constants.AmenityToilet,
		constants.AmenityTV,
	}
	amenities := make([]model.AmenityConstant, 0)
	for _, a := range commonAmenities {
		amenities = append(amenities, model.AmenityConstant{
			Value:       a.String(),
			DisplayName: a.GetDisplayName(),
		})
	}

	// Sort options
	sortOptions := []model.ConstantDisplay{
		{Value: "departure_time", DisplayName: "Giờ đi"},
		{Value: "price", DisplayName: "Giá vé"},
		{Value: "duration", DisplayName: "Thời gian hành trình"},
	}

	// Vietnam cities
	cities := []string{
		"Hà Nội",
		"TP. Hồ Chí Minh",
		"Đà Nẵng",
		"Đà Lạt",
		"Hải Phòng",
		"Cần Thơ",
		"Huế",
		"Nha Trang",
		"Quảng Ninh",
		"Bắc Ninh",
		"Bình Dương",
		"Bình Định",
		"Bình Phước",
		"Bến Tre",
		"Gia Lai",
		"Kiên Giang",
		"Lâm Đồng",
		"Long An",
		"Nam Định",
		"Nghệ An",
		"Phú Yên",
		"Quảng Bình",
		"Quảng Nam",
		"Quảng Ngãi",
		"Sóc Trăng",
		"Tây Ninh",
		"Thái Nguyên",
		"Thanh Hóa",
		"Tiền Giang",
		"Trà Vinh",
		"Vĩnh Long",
		"Vũng Tàu",
	}

	return &model.SearchFilterConstants{
		SortOptions: sortOptions,
		PriceRanges: priceRanges,
		TimeSlots:   timeSlots,
		SeatTypes:   seatTypes,
		Amenities:   amenities,
		Cities:      cities,
	}, nil
}

func (s *ConstantsServiceImpl) GetCities(ctx context.Context) ([]string, error) {
	return []string{
		"Hà Nội",
		"TP. Hồ Chí Minh",
		"Đà Nẵng",
		"Đà Lạt",
		"Hải Phòng",
		"Cần Thơ",
		"Huế",
		"Nha Trang",
		"Quảng Ninh",
		"Bắc Ninh",
		"Bình Dương",
		"Bình Định",
		"Bình Phước",
		"Bến Tre",
		"Gia Lai",
		"Kiên Giang",
		"Lâm Đồng",
		"Long An",
		"Nam Định",
		"Nghệ An",
		"Phú Yên",
		"Quảng Bình",
		"Quảng Nam",
		"Quảng Ngãi",
		"Sóc Trăng",
		"Tây Ninh",
		"Thái Nguyên",
		"Thanh Hóa",
		"Tiền Giang",
		"Trà Vinh",
		"Vĩnh Long",
		"Vũng Tàu",
	}, nil
}

func (s *ConstantsServiceImpl) GetAllConstants(ctx context.Context) (*model.ConstantsResponse, error) {
	busConstants, err := s.GetBusConstants(ctx)
	if err != nil {
		return nil, err
	}

	routeConstants, err := s.GetRouteConstants(ctx)
	if err != nil {
		return nil, err
	}

	tripConstants, err := s.GetTripConstants(ctx)
	if err != nil {
		return nil, err
	}

	searchFilters, err := s.GetSearchFilterConstants(ctx)
	if err != nil {
		return nil, err
	}

	return &model.ConstantsResponse{
		Bus:           *busConstants,
		Route:         *routeConstants,
		Trip:          *tripConstants,
		SearchFilters: *searchFilters,
	}, nil
}

File: frontend/lib/api/constants-service.ts
import apiClient, { ApiResponse, handleApiError } from "./client";

/**
 * Constant display value structure
 */
export interface ConstantDisplay {
  value: string;
  display_name: string;
}

/**
 * Seat type constant with pricing information
 */
export interface SeatTypeConstant {
  value: string;
  display_name: string;
  price_multiplier: number;
}

/**
 * Amenity constant
 */
export interface AmenityConstant {
  value: string;
  display_name: string;
}

/**
 * Bus type constant
 */
export interface BusTypeConstant {
  value: string;
  display_name: string;
}

/**
 * Stop type constant
 */
export interface StopTypeConstant {
  value: string;
  display_name: string;
}

/**
 * Trip status constant
 */
export interface TripStatusConstant {
  value: string;
  display_name: string;
}

/**
 * Filter price range
 */
export interface FilterPriceRange {
  min: number;
  max: number;
}

/**
 * Filter time slot
 */
export interface FilterTimeSlot {
  start_time: string; // HH:MM format
  end_time: string; // HH:MM format
  display_name: string;
}

/**
 * Search filter constants
 */
export interface SearchFilterConstants {
  sort_options: ConstantDisplay[];
  price_ranges: FilterPriceRange[];
  time_slots: FilterTimeSlot[];
  seat_types: SeatTypeConstant[];
  amenities: AmenityConstant[];
  cities: string[];
}

/**
 * Bus-related constants
 */
export interface BusConstants {
  seat_types: SeatTypeConstant[];
  amenities: AmenityConstant[];
  bus_types: BusTypeConstant[];
}

/**
 * Route-related constants
 */
export interface RouteConstants {
  stop_types: StopTypeConstant[];
}

/**
 * Trip-related constants
 */
export interface TripConstants {
  trip_statuses: TripStatusConstant[];
}

/**
 * All constants grouped by domain
 */
export interface ConstantsResponse {
  bus: BusConstants;
  route: RouteConstants;
  trip: TripConstants;
  search_filters: SearchFilterConstants;
}

/**
 * Valid constant type values for API query parameter
 */
export type ConstantType =
  | "bus"
  | "route"
  | "trip"
  | "search_filters"
  | "cities";

/**
 * Get all constants grouped by domain
 */
export const getAllConstants = async (): Promise<ConstantsResponse> => {
  try {
    const response = await apiClient.get<ApiResponse<ConstantsResponse>>(
      "/trip/api/v1/constants",
    );

    if (!response.data.data) {
      throw new Error(
        response.data.message ||
          response.data.error ||
          "Failed to get constants",
      );
    }

    return response.data.data;
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * Get bus-related constants (seat types, amenities, bus types)
 */
export const getBusConstants = async (): Promise<BusConstants> => {
  try {
    const response = await apiClient.get<ApiResponse<BusConstants>>(
      "/trip/api/v1/constants?type=bus",
    );

    if (!response.data.data) {
      throw new Error(
        response.data.message ||
          response.data.error ||
          "Failed to get bus constants",
      );
    }

    return response.data.data;
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * Get route-related constants (stop types)
 */
export const getRouteConstants = async (): Promise<RouteConstants> => {
  try {
    const response = await apiClient.get<ApiResponse<RouteConstants>>(
      "/trip/api/v1/constants?type=route",
    );

    if (!response.data.data) {
      throw new Error(
        response.data.message ||
          response.data.error ||
          "Failed to get route constants",
      );
    }

    return response.data.data;
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * Get trip-related constants (trip statuses)
 */
export const getTripConstants = async (): Promise<TripConstants> => {
  try {
    const response = await apiClient.get<ApiResponse<TripConstants>>(
      "/trip/api/v1/constants?type=trip",
    );

    if (!response.data.data) {
      throw new Error(
        response.data.message ||
          response.data.error ||
          "Failed to get trip constants",
      );
    }

    return response.data.data;
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * Get search filter constants (sort options, time slots, price ranges, etc.)
 */
export const getSearchFilterConstants =
  async (): Promise<SearchFilterConstants> => {
    try {
      const response = await apiClient.get<ApiResponse<SearchFilterConstants>>(
        "/trip/api/v1/constants?type=search_filters",
      );

      if (!response.data.data) {
        throw new Error(
          response.data.message ||
            response.data.error ||
            "Failed to get search filter constants",
        );
      }

      return response.data.data;
    } catch (error) {
      const errorMessage = handleApiError(error);
      throw new Error(errorMessage);
    }
  };

/**
 * Get list of cities
 */
export const getCities = async (): Promise<string[]> => {
  try {
    const response = await apiClient.get<ApiResponse<string[]>>(
      "/trip/api/v1/constants?type=cities",
    );

    if (!response.data.data) {
      throw new Error(
        response.data.message || response.data.error || "Failed to get cities",
      );
    }

    return response.data.data;
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};
